### DISCUSSION TITLE
AbsoluteBlockBatch - Unable to apply ChunkBatch to unloaded chunk

### DISCUSSION THREAD

**Discussion by muelleel** [Minestom/Minestom#1263] [Help] [answered]
Hi,

I'm trying to apply an AbsoluteBlockBatch on an Instance, while the Player is currently not on the Instance or is far away.

When applying I get a bunch of these messages:

`[Ms-TickScheduler] [11:47:58] (ChunkBatch.apply) - WARN - Unable to apply ChunkBatch to unloaded chunk (0, 0) in bf7fb101-1054-4e31-8d64-ca2bb09f052c.`

I already execute `batch.awaitReady();` before applying!

I'm aware of the `instance.loadChunk(x, y);` method, but what is the best way to load all Chunks the batch is affecting?

Do I have to implement the chunk loading manually or is there an better way to this?
Maybe even place blocks without loading the chunks?

(doing the placements with the `instance.setGenerator(generator)` is not an Option for my use case)

EDIT: I use the `instance.setGenerator(...)` for chunk generation.

Thanks

**Comment by MrGazdag** [comment, marked as answer]
Well, i'm afraid you will have to load the chunks manually using `instance.loadChunk(int, int)`.
There is an internal utility to load chunks in batch, which is `ChunkUtils.optionalLoadAll(Instance, long[], ChunkCallback)`. You put all the chunks you want in the long array (the longs themselves contain the position of the chunk encoded in the bits, you can use `ChunkUtils.getChunkIndex(chunkX, chunkZ)`) and you get a CompletableFuture, that will be completed when all chunks have been loaded. You could then use the chaining methods of the CompletableFuture to apply the batch right after the chunks have been loaded. 
Keep in mind that for this to work, you will need to know which chunks you will be affecting.

If you know two opposite corners of the affected area, you can do the following:
```java
Instance instance = ...;
Vec corner1 = ...;
Vec corner2 = ...;
AbsoluteBlockBatch batch = ...;

// Find the smallest and largest chunk coordinates
int startChunkX = Math.min(corner1.chunkX(), corner2.chunkX());
int startChunkZ = Math.min(corner1.chunkZ(), corner2.chunkZ());

int endChunkX = Math.max(corner1.chunkX(), corner2.chunkX());
int endChunkZ = Math.max(corner1.chunkZ(), corner2.chunkZ());

int xSize = endChunkX-startChunkX+1;
int zSize = endChunkZ-startChunkZ+1;

// Create the long array
long[] array = new long[xSize * zSize];
for (int chunkX = 0; chunkX < xSize; chunkX++) {
    for (int chunkZ = 0; chunkZ < zSize; chunkZ++) {
        array[chunkX*zSize + chunkZ] = ChunkUtils.getChunkIndex(chunkX, chunkZ);
    }
}
ChunkUtils.optionalLoadAll(instance, array, null).thenRun(() -> {
    // Chunks have completed loading
    batch.apply(instance),
});
```
The last part can even be replaced with this shorter version too:
```java
ChunkUtils.optionalLoadAll(instance, array, null).thenRun(() -> batch.apply(instance));
```

**Reply by muelleel** [reply]
Thanks for the hint to the ChunkUtils Class! Sounds good for the start.

> you will need to know which chunks you will be affecting

I have a bunch of for loops which use `batch.setBlock` to set Blocks in the Batch. Is there a way to get the affected Chunks from an batch?
something like `long[] list = batch.getAffectedChunkIndexes()`

That would be great. How would I do it without such an Method?

**Reply by MrGazdag** [reply]
I'm afraid there are no such methods for this, though it would be trivial to query from the current implementation of the AbsoluteBlockBatch. Nevertheless, I edited the original answer to contain some example code on how to load all chunks if you know two corners of the affected area. I might make a PR to make a method like this.

**Reply by muelleel** [reply]
Thank you,
the example will work fine for me!

> I might make a PR to make a method like this.

A PR for the `batch.getAffectedChunkIndexes()` imeplementation or for the affected area?

Would be great to have something in that direction implemented in Minestom :+1: 

Thanks!

**Reply by MrGazdag** [reply]
See #1264

**Reply by muelleel** [reply]
Hey,
since your PR looked so pretty I tried to squeeze that little function into the current Minestom version with reflection myself...
so if anyone has the same Problem here it is:

```
public class BatchUtil {
    public static long[] getAffectedChunks(AbsoluteBlockBatch batch) {
        try {
            Field field = batch.getClass().getDeclaredField("chunkBatchesMap");
            field.setAccessible(true);

            Long2ObjectMap<ChunkBatch> chunkBatchesMap = (Long2ObjectMap<ChunkBatch>) field.get(batch);

            return chunkBatchesMap.keySet().toLongArray();

        } catch (NoSuchFieldException | IllegalAccessException e) {
            throw new RuntimeException(e);
        }
    }
}

```

Usage:
```

Batch batch = ...
Instance instance = ...

//do stuff with setBlock on the batch...

long[] affectedChunks = BatchUtil.getAffectedChunks(batch);  //BatchUtil is the class from above

ChunkUtils.optionalLoadAll(instance, affectedChunks, null).thenRun(() -> {
    batch.apply(instance, () -> {
        // done
    });
});
```
